Modules/MUON/MCH/src/TrackPlotter.cxx

```diff
@@ -0,0 +1,356 @@
+      mNofClustersPerDualSampa->Fill(dsbinx(deId, dsId));
+    }
+    int chamberId = cluster.getChamberId();
+    mClusterSizePerChamber->Fill(chamberId + 1, cluster.nDigits);
+    std::set<PadSize> padSizes;
+    std::set<double, decltype(padPosCompare)> xpos(padPosCompare);
+    std::set<double, decltype(padPosCompare)> ypos(padPosCompare);
+    for (int dix = cluster.firstDigit; dix < cluster.firstDigit + cluster.nDigits; ++dix) {
+      const auto& digit = digits[dix];
+      auto padid = digit.getPadID();
+      PadSize padSize{ seg.padSizeX(padid),
+                       seg.padSizeY(padid) };
+      padSizes.emplace(padSize);
+      if (seg.isBendingPad(padid)) {
+        ypos.emplace(seg.padPositionY(padid));
+      } else {
+        xpos.emplace(seg.padPositionX(padid));
+      }
+    }
+    mClusterBendingSizePerChamber->Fill(chamberId + 1, ypos.size());
+    mClusterNonBendingSizePerChamber->Fill(chamberId + 1, xpos.size());
+  }
+}
+
+void TrackPlotter::fillTrackPairHistos(gsl::span<const o2::mch::TrackMCH> tracks)
+{
+  if (tracks.size() > 1) {
+    for (auto i = 0; i < tracks.size(); i++) {
+      auto ti = getMomentum4D(tracks[i]);
+      for (auto j = i + 1; j < tracks.size(); j++) {
+        if (tracks[i].getSign() == tracks[j].getSign()) {
+          continue;
+        }
+        auto tj = getMomentum4D(tracks[j]);
+        auto p = ti + tj;
+        mMinv->Fill(p.M());
+      }
+    }
+  }
+}
+
+bool TrackPlotter::fillTrackHistos(const o2::mch::TrackMCH& track,
+                                   gsl::span<const o2::mch::Cluster> clusters,
+                                   gsl::span<const o2::mch::Digit> digits)
+{
+  mNofClustersPerTrack->Fill(track.getNClusters());
+  mTrackChi2OverNDF->Fill(track.getChi2OverNDF());
+
+  const auto trackClusters = clusters.subspan(track.getFirstClusterIdx(), track.getNClusters());
+  fillClusterHistos(trackClusters, digits);
+
+  auto clustersPerChamber = getClustersPerChamber(trackClusters);
+
+  for (auto i = 0; i < clustersPerChamber.size(); i++) {
+    mNofClustersPerChamber->Fill(i + 1, clustersPerChamber[i]);
+  }
+
+  auto p = track.getP(); // uncorrected p
+
+  o2::mch::TrackParam trackParamAtDCA(track.getZ(), track.getParameters());
+  if (!o2::mch::TrackExtrap::extrapToVertexWithoutBranson(trackParamAtDCA, 0.0)) {
+    return false;
+  }
+
+  o2::mch::TrackParam trackParamAtOrigin(track.getZ(), track.getParameters());
+  if (!o2::mch::TrackExtrap::extrapToVertex(trackParamAtOrigin, 0.0, 0.0, 0.0, 0.0, 0.0)) {
+    return false;
+  }
+  double dcaX = trackParamAtDCA.getBendingCoor();
+  double dcaY = trackParamAtDCA.getNonBendingCoor();
+  double dca = std::sqrt(dcaX * dcaX + dcaY * dcaY);
+  mTrackDCA->Fill(dca);
+  mTrackPDCA->Fill(p * dca);
+
+  auto muon = getMomentum4D(trackParamAtOrigin);
+
+  mTrackEta->Fill(muon.eta());
+  mTrackPhi->Fill(muon.phi() * TMath::RadToDeg() + 180);
+  mTrackPt->Fill(muon.pt());
+
+  o2::mch::TrackExtrap::extrapToZ(trackParamAtOrigin, -505.);
```

Comment:
For Rabs extrapolation must be done from parameters at first cluster not from parameters already extrapolated to vextex.
Beside, possible extrapolation failure should also be catch to return false if it happens.

Reply:
Modified as suggested